src/realizations/catchment/Bmi_Multi_Formulation.cpp

```diff
@@ -0,0 +1,266 @@
+        std::string type_name = formulation_config.at("name").as_string();
+        std::string identifier = get_catchment_id() + "." + std::to_string(i);
+        std::shared_ptr<Bmi_Module_Formulation<bmi::Bmi>> module;
+        if (type_name == "bmi_c") {
+            module = std::dynamic_pointer_cast<Bmi_Module_Formulation<bmi::Bmi>>(std::make_shared<Bmi_C_Formulation>(identifier, forcing, output));
+        }
+        else {
+            throw runtime_error(get_formulation_type() + " received unexpected subtype formulation " + type_name);
+        }
+        modules[i] = module;
+        // Call create_formulation on each formulation
+        module->create_formulation(formulation_config.at("params").get_values());
+
+        // Set this up for placing in the module_variable_maps member variable
+        std::shared_ptr<std::map<std::string, std::string>> var_aliases;
+        var_aliases = std::make_shared<std::map<std::string, std::string>>(
+                std::map<std::string, std::string>());
+
+        for (const std::string &var_name : module->get_bmi_input_variables()) {
+            std::string framework_alias = module->get_config_mapped_variable_name(var_name);
+            (*var_aliases)[framework_alias] = var_name;
+            // If framework_name is not in collection from which we have available data sources ...
+            if (availableData.count(framework_alias) != 1) {
+                throw std::runtime_error(
+                        "Multi BMI cannot be created with module " + module->get_model_type_name() + " with input " +
+                        "variable " + framework_alias +
+                        (var_name == framework_alias ? "" : " (an alias of BMI variable " + var_name + ")") +
+                        " when there is no previously-enabled source for this input.");
+            }
+            else {
+                module->input_forcing_providers[var_name] = availableData[framework_alias];
+                module->input_forcing_providers[framework_alias] = availableData[framework_alias];
+            }
+        }
+        // Also add the output variable aliases
+        for (const std::string &var_name : module->get_bmi_output_variables()) {
+            std::string framework_alias = module->get_config_mapped_variable_name(var_name);
+            (*var_aliases)[framework_alias] = var_name;
+            if (availableData.count(framework_alias) > 0) {
+                throw std::runtime_error(
+                        "Multi BMI cannot be created with module " + module->get_model_type_name() +
+                        " with output variable " + framework_alias +
+                        (var_name == framework_alias ? "" : " (an alias of BMI variable " + var_name + ")") +
+                        " because a previous module is using this output variable name/alias.");
+            }
+            availableData[framework_alias] = module;
+        }
+        module_variable_maps[i] = var_aliases;
+    }
```

Comment:
For these long, nested loops, it is often really helpful to put a comment after the closing bracket to indicate which loop has ended, i.e. `} // end for sub_formulations_list` or something to help clearly identify it.